### PR TITLE
Set updated tmax value to bidder's MakeRequest implementation

### PR DESCRIPTION
**Pull request by onkarvhanumante**
Reducing the amount of time bidders have to compensate for the processing time used by PBS to,
- fetch a stored request (if needed)
-  validate the OpenRTB request and split it into multiple requests sanitized for each bidder. 
-  time needed by PBS to prepare the auction response

### CODE REVIEW

File: exchange/bidder.go [hunk 1]
```diff
@@ -718,3 +724,28 @@ func getBidTypeForAdjustments(bidType openrtb_ext.BidType, impID string, imp []o
...
+}
+func (b *bidderTmaxCtx) Deadline() (deadline time.Time, ok bool) {
+	deadline, ok = b.ctx.Deadline()
+	return
+}
+
+func getBidderTmax(ctx bidderTmaxContext, requestTmaxMS int64, tmaxAdjustments config.TmaxAdjustments) int64 {
+	if tmaxAdjustments.Enabled &&
+		tmaxAdjustments.BidderResponseDurationMin != 0 &&
+		(tmaxAdjustments.BidderNetworkLatencyBuffer != 0 || tmaxAdjustments.PBSResponsePreparationDuration != 0) {
```
Comment (VeronikaSolovei9) [inline exchange/bidder.go:745]: This block of code looks like a good candidate to be evaluated once at the server start up. These parameters come with the config and will not change per-request.

For the `exchange.AuctionRequest` you may use another, more convenient structure with some pre-processed values from configs.

Another point toward this idea is that you use same exact structure you used for config representation. 
These structs should be separated (model-view-presenter design pattern). Look at the `Activities` example in `exchange.AuctionRequest`.

**Review by VeronikaSolovei9 — COMMENTED** [review]
Added comments related to code optimization and better structuring. 

Just want to confirm: 
-do duration values in `type TmaxAdjustments struct` should be validated? 
-do duration values in `type TmaxAdjustments struct` should be converted to Duration?
-will these values somewhow interact with timeout calculated earlier at the endpoint start?
```go
timeout := deps.cfg.AuctionTimeouts.LimitAuctionTimeout(time.Duration(req.TMax) * time.Millisecond)
	if timeout > 0 {
		var cancel context.CancelFunc
		ctx, cancel = context.WithDeadline(ctx, start.Add(timeout))
		defer cancel()
	}
```

File: exchange/bidder.go (hunk 1 above)
Comment (onkarvhanumante) [reply exchange/bidder.go:745]: >This block of code looks like a good candidate to be evaluated once at the server start up. These parameters come with the config and will not change per-request.

`0` is default value   for `BidderResponseDurationMin, .BidderNetworkLatencyBuffer and .PBSResponsePreparationDuration`.  So we cannot add them at server startup or validate them to return error if 0.  

By checking these values `!=0`, we make sure that deadline from context are extracted and reduced only if needed. This saves us with some time

File: endpoints/openrtb2/amp_auction.go [hunk 2]
```diff
@@ -248,6 +248,7 @@ func (deps *endpointDeps) AmpAuction(w http.ResponseWriter, r *http.Request, _ h
 		QueryParams:                r.URL.Query(),
 		TCF2Config:                 tcf2Config,
 		Activities:                 activities,
+		TmaxAdjustments:            &deps.cfg.TmaxAdjustments,
```
[the remark this answers is not in the document]
Comment (onkarvhanumante) [reply endpoints/openrtb2/amp_auction.go:251]: `TmaxAdjustments` is being passed/ planned to passed to various downstream functions and these functions will have access to `tmaxAdjustments.Enabled` field.  

Also consider a scenario, if in future if following code was changed. Then if downstream function don't have `tmaxAdjustments.Enabled` check then behaviour of downstream function might change

```
tmax := &deps.cfg.TmaxAdjustments 
if !deps.cfg.TmaxAdjustments.Enabled { 
  tmax = nil 
}
```

**Comment by onkarvhanumante** [comment]
@VeronikaSolovei9  for this https://github.com/prebid/prebid-server/pull/2930#pullrequestreview-1538422600,



>  -do duration values in type TmaxAdjustments struct should be validated?

Zero is the default values of TmaxAdjustments fields.  Also data type of TmaxAdjustments fields is uint which implies values can't be less than 0 and marshalling of config (pbs.yml) at start up would failed. So no additional validation is needed


>  -do duration values in type TmaxAdjustments struct should be converted to Duration?
    
these values are duration (amount of time) specified in ms  
 https://github.com/prebid/prebid-server/blob/23fffa11d0b832ed226a1ded70e3dcfd799c500f/config/config.go#L1545-L1562



>  -will these values somewhow interact with timeout calculated earlier at the endpoint start?
       ```
         timeout := deps.cfg.AuctionTimeouts.LimitAuctionTimeout(time.Duration(req.TMax) * time.Millisecond)
	if timeout > 0 {
		var cancel context.CancelFunc
		ctx, cancel = context.WithDeadline(ctx, start.Add(timeout))
		defer cancel()
	}
      ```

yes timeout deadline earlier at the endpoint start is used further to get elapsed time for determining remaining duration

File: exchange/bidder.go (hunk 1 above)
Comment (VeronikaSolovei9) [reply exchange/bidder.go:745]: Sorry, this is not what I mean, I didn't express myself clear. My general proposal here is to preprocess data from deps.cfg.TmaxAdjustments and save it to a new structure that will be passed to the endpoints.
Here is a draft with changes: https://github.com/VeronikaSolovei9/prebid-server/pull/4 
Please see my commit on top of your changes: https://github.com/VeronikaSolovei9/prebid-server/pull/4/commits/99653260ba2f5a1b378a46dfb0bfebe424f18f9a

**Comment by VeronikaSolovei9** [comment]
I tried to test it locally with negative values. You are right, uint cannot be negative. Here is my pbs.json and unmarshalled values in config:
Screenshot 2023-07-20 at 11 37 12 AM

Also do you think sum of all durations in tmaxAdjustment should not exceed general request timeout?

File: exchange/bidder.go (hunk 1 above)
Comment (onkarvhanumante) [reply exchange/bidder.go:745]: @VeronikaSolovei9 thanks for the draft. For now I will defer to using `config.TmaxAdjustments` struct. Lets build out this feature and then we can do this enhancement.  Being said that, lets see what other reviewers has feedback on this

Comment (VeronikaSolovei9) [reply exchange/bidder.go:745]: Why do you want to defer it? This change is simple and makes sense, it is more optimal than current approach and it's all done in draft so you see how it works already. It's better to do everything right at the beginning (especially simple things like this), rather then do it in the follow up PR that can possibly never happen or will be forgotten.

Comment (onkarvhanumante) [reply exchange/bidder.go]: @VeronikaSolovei9  https://github.com/prebid/prebid-server/pull/2930/commits/fb19f5c1af6a39a55aa6c564f1fac7e4802908e0 makes the changes

**Comment by VeronikaSolovei9** [comment]
We discussed the validation question offline. It will be possibly added in future PRs due to the team decision that validation is optional here.

**Review by VeronikaSolovei9 — APPROVED** [review]
Thank you for the changes, LGTM

**Pull request merged**